XRTK.SpatialPersistence.ASA/Packages/com.xrtk.spatial-persistence.asa/Runtime/Providers/ASASpatialPersistenceDataProvider.cs

```diff
@@ -178,32 +178,52 @@ public async void StopSpatialPersistenceProvider()
         /// <param name="args"></param>
         private void CloudManager_AnchorLocated(object sender, AnchorLocatedEventArgs args)
         {
-            if (Guid.TryParse(args.Identifier, out var anchorGuid))
+            try
             {
-                if (!detectedAnchors.ContainsKey(anchorGuid))
+                if (Guid.TryParse(args.Identifier, out var anchorGuid))
                 {
-                    detectedAnchors.Add(anchorGuid, args.Anchor);
-                }
+                    // If an anchor is found but has no Anchor data, create a new CloudSpatialAnchor
+                    if (!detectedAnchors.ContainsKey(anchorGuid))
+                    {
+                        detectedAnchors.Add(anchorGuid, args.Anchor);
+                    }
 
-                //Android and iOS require coordinate from stored Anchor
+                    // Android and iOS require coordinate from stored Anchor
 #if UNITY_ANDROID || UNITY_IOS
-                var detectedAnchorPose = detectedAnchors[anchorGuid].GetPose();
+                    var detectedAnchorPose = detectedAnchors[anchorGuid].GetPose();
 #else
-                var detectedAnchorPose = Pose.identity;
+                    var detectedAnchorPose = Pose.identity;
 #endif
-                var anchoredObject = new GameObject($"Anchor - [{anchorGuid}]");
-                anchoredObject.transform.SetPositionAndRotation(detectedAnchorPose.position, detectedAnchorPose.rotation);
 
-                var attachedAnchor = anchoredObject.EnsureComponent<CloudNativeAnchor>();
-                attachedAnchor.CloudToNative(detectedAnchors[anchorGuid]);
+                    var anchoredObject = new GameObject($"Anchor - [{anchorGuid}]");
+                    anchoredObject.transform.SetPositionAndRotation(detectedAnchorPose.position, detectedAnchorPose.rotation);
+
+                    var attachedAnchor = anchoredObject.EnsureComponent<CloudNativeAnchor>();
+                    attachedAnchor.CloudToNative(detectedAnchors[anchorGuid]);
 
-                AnchorLocated?.Invoke(anchorGuid, anchoredObject);
+                    AnchorLocated?.Invoke(anchorGuid, anchoredObject);
+                }
+                else
+                {
+                    var errorMessage = $"Anchor returned from service but Identifier was invalid [{args.Identifier}]";
+                    Debug.LogError(errorMessage);
```

Comment:
Do we also need to log the error if we're invoking an error anyway?

This whole patter seems really out of place since raising an exception would make more sense

Reply:
The reason the Android app is crashing is because the ASA service is raising a NullReferenceException, and lord knows what else.
If you feel the exception data would be useful, we can certainly update it to expose it.

Reply:
I just find this to be an antipattern is all. If we're already raising an event about the error what difference does it make if we just raise an exception?

Reply:
I find it an anti-pattern throwing an exception when an exception has been thrown that is crashing the device.
Just throwing an exception would not help as it wouldn't report back the GUI of the anchor which is faulty and the client application needs to handle.
The event is to pass control back to the client application to deal with that action. An exception would do the same but wouldn't inform the client application of the data that needs actioning.

Can you propose an alternate solution that handles this use case instead of this back and to?

Reply:
> Can you propose an alternate solution that handles this use case instead of this back and to?

Throwing an exception.

> pass control back to the client application to deal with that action
That's why try/catches exist in the first place.

redundant action is redundant. You're effectively duplicating something that already exists but also taking control away from the client as they can't disable the debug message either, but you're still passing the error message back to them anyway. It's like you're half way in between handling it for them or just letting it fail.

Reply:
Either that or just handle the exception in your client like you should be doing anyway. I still don't like the way these events work in this system.